python/feathub/feature_views/feature_view.py

```diff
@@ -192,3 +214,50 @@ def get_bounded_view(self) -> TableDescriptor:
         feature_view = deepcopy(self)
         feature_view.source = self.get_resolved_source().get_bounded_view()
         return feature_view
+
+    def _derive_feature_dtype(self, feature: Feature) -> Optional[DType]:
+        transform = feature.transform
+        if isinstance(transform, ExpressionTransform):
+            return self._derive_feature_dtype_from_feathub_expression(transform.expr)
+        elif isinstance(transform, OverWindowTransform):
+            expr_result_type = self._derive_feature_dtype_from_feathub_expression(
+                transform.expr
+            )
+            return transform.agg_func.get_result_type(expr_result_type)
+        elif isinstance(transform, SlidingWindowTransform):
+            expr_result_type = self._derive_feature_dtype_from_feathub_expression(
+                transform.expr
+            )
+            return transform.agg_func.get_result_type(expr_result_type)
+        elif isinstance(transform, JoinTransform):
+            # We don't need to derive data type of the feature to join. Its data type is
+            # defined by the Feature to join.
+            return None
+        else:
+            raise FeathubException(
+                f"Cannot derive feature data type of {type(transform)} feature"
+            )
+
+    def _derive_feature_dtype_from_feathub_expression(self, expr: str) -> DType:
+        ast = feathub_expr_parser.parse(expr)
+        variable_types = {}
+
+        for input_feature_name in ast.get_variables():
```

Comment:
Instead of using `TableDescriptor#get_feature` to get the dtype for variables used in the given expression, an alternative approach is to get the dtype for all features defined on `source` and this FeatureView, and re-use this method to determine the dtype for arbitrary expressions used in this FeatureView.

The latter approach seems simpler since the same mapping can be repeatedly re-used and we don't need to add the abstract method `ExprAST#get_variables`. We can add `TableDescriptor#get_output_fields(self)` to expose this information.

What do you think?